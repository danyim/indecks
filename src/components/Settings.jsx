import PropTypes from 'prop-types'
import React from 'react'
import ExportDeckButton from './ExportDeckButton'
import styles from '../styles/components/Settings.styl'

const propTypes = {
  decks: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    cards: PropTypes.arrayOf(PropTypes.shape({
      title: PropTypes.string.isRequired,
      answer: PropTypes.string.isRequired,
      index: PropTypes.number.isRequired
    }).isRequired).isRequired
  }).isRequired).isRequired,
  deckCount: PropTypes.number.isRequired,
  removeAllDecks: PropTypes.func.isRequired
}

// articles: PropTypes.arrayOf(PropTypes.shape({
//   url: PropTypes.string.isRequired,
//   title: PropTypes.string.isRequired,
//   author: PropTypes.string.isRequired,
// }).isRequired).isRequired,

const defaultProps = {}

class Settings extends React.Component {
  constructor (props) {
    super(props)
    this.removeAllDecks = this.removeAllDecks.bind(this)
  }

  removeAllDecks (count) {
    if (confirm(`Are you sure you want to delete all ${count} decks? ` +
      'This action is permanent.')) {
      this.props.removeAllDecks()
    }
  }

  renderDeleteAll (deckCount) {
    return deckCount === 0
      ? (
        <button
          className='btn-delete'
          disabled='disabled'
        >
          Delete all decks from local storage
        </button>
      )
      : (
        <button
          className='btn-delete'
          onClick={() => this.removeAllDecks(deckCount)}
        >
          Delete all {deckCount} deck(s) from local storage
        </button>
      )
  }

  render () {
    const { deckCount } = this.props

    return (
      <section className={`${styles.settings}`}>
        <h2 className={`${styles.header}`}>Settings</h2>
        <div className={`${styles['settings-content']}`}>
          <p>
            The decks you create are automatically saved to your browser&apos;s
            local storage.
          </p>
          <ExportDeckButton
            filename='indecks.json'
            label='Download your deck collection as JSON'
            className='button'
            disabled={deckCount === 0}
            style={{}}
            exportFile={() => JSON.stringify(this.props.decks, null, 2)}
          />
          {this.renderDeleteAll(deckCount)}
        </div>
      </section>
    )
  }
}

Settings.propTypes = propTypes
Settings.defaultProps = defaultProps

export default Settings
