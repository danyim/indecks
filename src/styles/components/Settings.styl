@import '../_variables';

.grid-figure

h2.header
  margin-top: 0
  padding-bottom: .3rem
  border-bottom: .6rem solid black

.settings-view
  max-width: pageWidth

.settings
  // width: 100%

.settings-content
  //

.setings-bottom
  align-self: flex-end
