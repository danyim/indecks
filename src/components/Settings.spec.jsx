import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme'
import Settings from './Settings'

const defaultProps = {
  decks: [{
    id: 'ABCDEFG',
    title: 'Sample Deck',
    description: 'Deck description',
    cards: []
  }],
  deckCount: 1,
  removeAllDecks: () => {}
}

function setup (props = defaultProps) {
  const wrapper = shallow(<Settings {...props} />)

  return {
    props,
    wrapper
  }
}

function setupFull (props = defaultProps) {
  const wrapper = mount(<Settings {...props} />)

  return {
    props,
    wrapper
  }
}

describe('Settings', () => {
  it('should render self and subcomponents', () => {
    const tree = renderer.create(
      <Settings {...defaultProps}>Hello</Settings>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render the export button', () => {
    const { wrapper } = setupFull()
    expect(wrapper.find('ExportDeckButton').exists()).toBe(true)
  })

  it('should render the "delete all from local storage" button with the correct number', () => {
    const deckCount = 5
    const { wrapper } = setup({
      ...defaultProps,
      deckCount
    })

    const button = wrapper.findWhere(n => n.text() === `Delete all ${deckCount} deck(s) from local storage`)
    expect(button.length).toBe(1)
  })

  it('should render a disabled "delete all from local storage" button if no decks are found', () => {
    const deckCount = 0
    const { wrapper } = setup({
      ...defaultProps,
      deckCount
    })

    const button = wrapper.find('button.btn-delete[children="Delete all decks from local storage"]')
    expect(button.exists()).toBe(true)
    expect(button.prop('disabled')).toBe('disabled')
  })

  it('should call the removeAllDecks prop when button is clicked', () => {
    spyOn(window, 'confirm').and.returnValue(true)
    const handler = jest.fn()
    const { wrapper } = setup({
      ...defaultProps,
      removeAllDecks: handler
    })

    const button = wrapper.find('button.btn-delete')
    button.simulate('click')
    expect(handler.mock.calls.length).toBe(1)
  })

  it('should convert the deck into JSON when the export button is clicked', () => {
    const { props, wrapper } = setupFull()

    const exportButton = wrapper.find('ExportDeckButton')
    expect(exportButton.prop('exportFile')()).toEqual(JSON.stringify(props.decks, null, 2))
  })
})
