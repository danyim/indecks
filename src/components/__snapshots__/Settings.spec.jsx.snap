// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings should render self and subcomponents 1`] = `
<section
  className="settings"
>
  <h2
    className="header"
  >
    Settings
  </h2>
  <div
    className="settings-content"
  >
    <p>
      The decks you create are automatically saved to your browser's local storage.
    </p>
    <Connect(AuthContainer) />
    <ExportDeckButton
      className="button"
      disabled={false}
      exportFile={[Function]}
      filename="indecks.json"
      label="Download your deck collection as JSON"
      style={Object {}}
    />
    <button
      className="btn-delete"
      onClick={[Function]}
    >
      Delete all 
      1
       deck(s) from local storage
    </button>
  </div>
</section>
`;
